File: opencv-python/codeferm/peopledetect.py
"""
Copyright (c) Steven P. Goldsmith. All rights reserved.

Created by Steven P. Goldsmith on December 24, 2013
sgjava@gmail.com
"""

"""Histogram of Oriented Gradients ([Dalal2005]) object detector.

sys.argv[1] = source file or will default to "../../resources/960x720.mp4" if no args passed.

@author: sgoldsmith

"""
# Configure logger

import logging, sys, time, cv2

logger = logging.getLogger("PeopleDetect")
logger.setLevel("INFO")
formatter = logging.Formatter("%(asctime)s %(levelname)-8s %(module)s %(message)s")
handler = logging.StreamHandler(sys.stdout)
handler.setFormatter(formatter)
logger.addHandler(handler)
# If no args passed then default to internal file
if len(sys.argv) < 2:
    url = "../../resources/walking.mp4"
else:
    url = sys.argv[1]
outputFile = "../../output/people-detect-python.avi"
videoCapture = cv2.VideoCapture(url)
logger.info("OpenCV %s" % cv2.__version__)
logger.info("Input file: %s" % url)
logger.info("Output file: %s" % outputFile)
logger.info("Resolution: %dx%d" % (videoCapture.get(cv2.CAP_PROP_FRAME_WIDTH),
                               videoCapture.get(cv2.CAP_PROP_FRAME_HEIGHT)))
hog = cv2.HOGDescriptor()
hog.setSVMDetector(cv2.HOGDescriptor_getDefaultPeopleDetector())
videoWriter = cv2.VideoWriter(outputFile, cv2.VideoWriter_fourcc(*'X264'), videoCapture.get(cv2.CAP_PROP_FPS),
                              (int(videoCapture.get(cv2.CAP_PROP_FRAME_WIDTH)), int(videoCapture.get(cv2.CAP_PROP_FRAME_HEIGHT))), True)
lastFrame = False
frames = 0
framesWithPeople = 0
start = time.time()
while not lastFrame:
    ret, image = videoCapture.read()
    if ret:
        foundLocations, foundWeights = hog.detectMultiScale(image, winStride=(8, 8), padding=(32, 32), scale=1.05)
        if len (foundLocations) > 0:
            framesWithPeople += 1
            i = 0
            for x, y, w, h in foundLocations:
                # Draw rectangle around fond object
                cv2.rectangle(image, (x, y), (x + w, y + h), (0, 255, 0), 2)
                # Print weight
                cv2.putText(image, "%1.2f" % foundWeights[i], (x, y - 4), cv2.FONT_HERSHEY_PLAIN, 1.5, (255, 255, 255), thickness=2, lineType=cv2.LINE_AA)
                i += 1
    else:
        lastFrame = True
    videoWriter.write(image)
    frames += 1

elapsed = time.time() - start
logger.info("%d frames, %d frames with people" % (frames, framesWithPeople))
logger.info("%4.1f FPS, elapsed time: %4.2f seconds" % (frames / elapsed, elapsed))
del videoCapture
del videoWriter

File: opencv-cpp/peopledetect/src/PeopleDetect.cpp
/*
 * Copyright (c) Steven P. Goldsmith. All rights reserved.
 *
 * Created by Steven P. Goldsmith on November 5, 2015
 * sgoldsmith@codeferm.com
 */

#include <iostream>
#include <sys/time.h>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

/**
 * Histogram of Oriented Gradients ([Dalal2005]) object detector.
 *
 * argv[1] = source file or will default to "../../resources/walking.mp4" if no
 * args passed.
 *
 * @author sgoldsmith
 * @version 1.0.0
 * @since 1.0.0
 */
int main(int argc, char *argv[]) {
	int return_val = 0;
	string url = "../../resources/walking.mp4";
	string output_file = "../../output/people-detect-cpp.avi";
	cout << CV_VERSION << endl;
	cout << "Press [Esc] to exit" << endl;
	VideoCapture capture;
	Mat image;
	// See if URL arg passed
	if (argc == 2) {
		url = argv[1];
	}
	cout << "Input file:" << url << endl;
	cout << "Output file:" << output_file << endl;
	capture.open(url);
	// See if video capture opened
	if (capture.isOpened()) {
		cout << "Resolution: " << capture.get(CAP_PROP_FRAME_WIDTH) << "x"
				<< capture.get(CAP_PROP_FRAME_HEIGHT) << endl;
		bool exit_loop = false;
		// Video writer
		VideoWriter writer(output_file, (int) capture.get(CAP_PROP_FOURCC),
				(int) capture.get(CAP_PROP_FPS),
				Size((int) capture.get(CAP_PROP_FRAME_WIDTH),
						(int) capture.get(CAP_PROP_FRAME_HEIGHT)));
		int frames = 0;
		int frames_with_people = 0;
		Scalar rect_color = Scalar(0, 255, 0);
		Scalar font_color = Scalar(255, 255, 255);
		Size win_stride = Size(8, 8);
		Size padding = Size(32, 32);
		Point font_point;
		HOGDescriptor hog;
		hog.setSVMDetector(cv::HOGDescriptor::getDefaultPeopleDetector());
		vector<Rect> found_locations;
		vector<double> found_weights;
		char weight[20];
		timeval start_time;
		gettimeofday(&start_time, 0);
		// Process all frames
		while (capture.read(image) && !exit_loop) {
			if (!image.empty()) {
				hog.detectMultiScale(image, found_locations, found_weights, 0.0,
						win_stride, padding, 1.05, 2.0, false);
				if (!found_locations.empty()) {
					frames_with_people++;
					// Mark all found locations
					for (size_t i = 0, max = found_locations.size(); i != max;
							++i) {
						rectangle(image, found_locations[i], rect_color, 2, 8,
								0);
						sprintf(weight, "%1.2f", found_weights[i]);
						font_point.x = found_locations[i].x;
						font_point.y = found_locations[i].y - 4;
						putText(image, weight, font_point, FONT_HERSHEY_PLAIN,
								1.5, font_color, 2, LINE_AA, false);
					}
				}
				// Write frame
				writer.write(image);
				frames++;
			} else {
				cout << "No frame captured" << endl;
				exit_loop = true;
			}
		}
		timeval end_time;
		gettimeofday(&end_time, 0);
		cout << frames << " frames, " << frames_with_people
				<< " frames with people" << endl;
		cout << "FPS " << (frames / (end_time.tv_sec - start_time.tv_sec))
				<< ", elapsed time: " << (end_time.tv_sec - start_time.tv_sec)
				<< " seconds" << endl;
		// Release VideoWriter
		writer.release();
		// Release VideoCapture
		capture.release();
	} else {
		cout << "Unable to open device" << endl;
		return_val = -1;
	}
	return return_val;
}
